### PR TITLE
Recreate canary-controller with generalization

### PR DESCRIPTION
**Pull request by potsbo**
## Why

We have https://github.com/wantedly/canary-controller, which can
- copy deployment
- inject an env var `CANARY_ENABLED=1`
- set hostname to `canary`
- set labels `canary: true` for `.metadata.labels`, `.spec.template.labels`, and `.spec.selector.matchLabels.`

But now I have another motivation
- copy deployment
- inject multiple env vars
- set hostname to any string
- set any labels for `.metadata.labels`, `.spec.template.labels`, and `.spec.selector.matchLabels.`
- set annotations for `.metadata.annotations`

These two are similar, which made me think that we should create another controller that can duplicate deployment that can be used for canary deploy or any other purposes

Also, canary-controller has been built with a kind of old technology for 2020
- using dep
- using kubebuilder v1
- using go 1.10.x

## What

Create a generalized version of canary-controller

Any feature available with canary-controller should be also available with this new controller

<details>

<summary>example</summary>

This Canary resource should be converted into the latter yaml
```yaml
apiVersion: canary.k8s.wantedly.com/v1beta1
kind: Canary
metadata:
  labels:
    controller-tools.k8s.io: "1.0"
  name: canary-sample
spec:
  targetDeploymentName: foo
  targetContainers: 
    - name: nginx
      image: nginx:latest
    - name: redis
      image: redis:5.0.5
```

```yaml
apiVersion: duplication.k8s.wantedly.com/v1beta1
kind: DeploymentCopy
metadata:
  name: deploymentcopy-sample
spec:
  targetDeploymentName: foo
  customLabels:
    canary: "true"
  nameSuffix: "canary"
  hostname: "hostname"
  targetContainers:
    - name: nginx
      image: nginx:latest
      env:
      - name: CANARY_ENABLED
        value: "1"
    - name: redis
      image: redis:5.0.5
      env:
      - name: CANARY_ENABLED
        value: "1"
```

</details>

### CODE REVIEW

File: api/v1beta1/deploymentcopy_types.go [hunk 1]
```diff
@@ -25,11 +26,33 @@ import (
...
+
+	// annotations in `CustomAnnotations` and those of `TargetDeploymentName` will be merged.
+	// When both have same keys, values in `Labels` will be applied
+	CustomAnnotations map[string]string `json:"customAnnotations,omitempty"`
+
+	// name defined in `TargetDeploymentName` will be copied
+	TargetDeploymentName string `json:"targetDeploymentName"`
+
+	// the copied deployment name will be `<TargetDeploymentName>-<NameSuffix>`
+	NameSuffix string `json:"nameSuffix"`
```
Comment (munisystem) [inline api/v1beta1/deploymentcopy_types.go:42]: In most cases with k8s ecosystem, the name of the resource generated by other resources is decided to the parent  (e.g. Deployment, Endpoint) and I think this approach looks good.
Do we need this field?

Comment (potsbo) [reply api/v1beta1/deploymentcopy_types.go:42]: You mean, for example, when you copy `foo` the copied deployment will be named like `foo-vxc6j`. 
I just wanted to make sure we can provide the same behavior that `canary-controller` does.
So as long as we are good to forget the behavior I also think we can drop this field.

Do you think it's good to go?

Comment (potsbo) [reply api/v1beta1/deploymentcopy_types.go:42]: or maybe we can use `.metadata.name` for the suffix.

Comment (munisystem) [reply api/v1beta1/deploymentcopy_types.go:42]: I think using `.metadata.name` is a better way.

**Pull request merged**